diff --git a/lib/index.js b/lib/index.js
index a65754e429f5da4f3ff7689808bb620d5967eea1..8497a7e8c62365082fc041b04e085cf2b0ef670d 100644
--- a/lib/index.js
+++ b/lib/index.js
@@ -77,7 +77,9 @@ export function initAuth(config, onLazyLoad // To set the default env vars
             return getSession(new Headers(request.headers), _config).then(async (authResponse) => {
                 const auth = await authResponse.json();
                 for (const cookie of authResponse.headers.getSetCookie())
-                    if ("headers" in response)
+                    if("setHeader" in response)
+                        response.setHeader("set-cookie", cookie);
+                    else if ("headers" in response)
                         response.headers.append("set-cookie", cookie);
                     else
                         response.appendHeader("set-cookie", cookie);
@@ -116,7 +118,9 @@ export function initAuth(config, onLazyLoad // To set the default env vars
         new Headers(request.headers), config).then(async (authResponse) => {
             const auth = await authResponse.json();
             for (const cookie of authResponse.headers.getSetCookie())
-                if ("headers" in response)
+                if("setHeader" in response)
+                    response.setHeader("set-cookie", cookie);
+                else if ("headers" in response)
                     response.headers.append("set-cookie", cookie);
                 else
                     response.appendHeader("set-cookie", cookie);
@@ -166,7 +170,12 @@ async function handleAuth(args, config, userMiddlewareOrRoute) {
     const finalResponse = new Response(response?.body, response);
     // Preserve cookies from the session response
     for (const cookie of sessionResponse.headers.getSetCookie())
-        finalResponse.headers.append("set-cookie", cookie);
+        if("setHeader" in finalResponse)
+            finalResponse.setHeader("set-cookie", cookie);
+        else if ("headers" in finalResponse)
+            finalResponse.headers.append("set-cookie", cookie);
+        else
+            finalResponse.appendHeader("set-cookie", cookie);
     return finalResponse;
 }
 function isSameAuthAction(requestPath, redirectPath, config) {
